interop-testing/src/main/java/io/grpc/testing/integration/AltsHandshakerTestService.java

```diff
@@ -53,48 +53,50 @@
       @Override
       public void onNext(HandshakerReq value) {
         log.log(Level.FINE, "request received: " + value);
-        switch (expectState) {
-          case CLIENT_INIT:
-            checkState(CLIENT_START.equals(value.getReqOneofCase()));
-            HandshakerResp initClient = HandshakerResp.newBuilder()
-                .setOutFrames(fakeOutput)
-                .build();
-            log.log(Level.FINE, "init client response " + initClient);
-            responseObserver.onNext(initClient);
-            expectState = State.SERVER_INIT;
-            break;
-          case SERVER_INIT:
-            checkState(SERVER_START.equals(value.getReqOneofCase()));
-            HandshakerResp initServer = HandshakerResp.newBuilder()
+        synchronized (this) {
```

Comment:
How did the race happen? Seems there could be a bug if the race does happen.

Reply:
It happens because the `state` change here is not atomic, e.g. at `responseObserver.onNext(resp)` is invoked before we move the state to the next, but this allows the server to send the final request. The final request is expecting `SERVER_FINISH` but because I didn't synchronized it, the state was wrong and response messed up. The [debug log](https://travis-ci.org/github/YifeiZhuang/grpc-java/jobs/757862871) seems telling about this and it happened pretty often.